modules/numbers/shared/src/main/scala/io/circe/numbers/BiggerDecimal.scala

```diff
@@ -263,132 +262,115 @@ object BiggerDecimal {
       var decIndex = -1
       var expIndex = -1
       var i = if (input.charAt(0) == '-') 1 else 0
-
-      var state =
-        if (i >= len) FAILED
-        else {
-          if (input.charAt(i) != '0') START
-          else {
-            i = i + 1
-            AFTER_ZERO
+      var parsedNonLeadingDigit: Boolean = false
+
+      if (i >= len) {
+        null
```

Comment:
I don't mind the short cut here, but we need to make it more clear this is a FAILED state.
```suggestion
        null // state = FAILED
```

Reply:
Comment added, FYI that is also listed in the ScalaDoc. 

I don't love using `null` here. I don't think the single boxing of one `Either` is a problematic cost, but this is most similar to the existing API.